x/group/keeper/grpc_query.go

```diff
@@ -305,54 +305,39 @@ func (q Keeper) getVotesByVoter(ctx sdk.Context, voter sdk.AccAddress, pageReque
 	return q.voteByVoterIndex.GetPaginated(ctx.KVStore(q.key), voter.Bytes(), pageRequest)
 }
 
-// Tally is a function that tallies a proposal by iterating through its votes,
-// and returns the tally result without modifying the proposal or any state.
-// TODO Merge with https://github.com/cosmos/cosmos-sdk/issues/11151
-func (q Keeper) Tally(ctx sdk.Context, p group.Proposal, groupId uint64) (group.TallyResult, error) {
-	// If proposal has already been tallied and updated, then its status is
-	// closed, in which case we just return the previously stored result.
-	if p.Status == group.PROPOSAL_STATUS_CLOSED {
-		return p.FinalTallyResult, nil
-	}
+func (q Keeper) TallyResult(goCtx context.Context, request *group.QueryTallyResultRequest) (*group.QueryTallyResultResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	proposalId := request.ProposalId
 
-	it, err := q.voteByProposalIndex.Get(ctx.KVStore(q.key), p.Id)
+	proposal, err := q.getProposal(ctx, proposalId)
 	if err != nil {
-		return group.TallyResult{}, err
+		return nil, err
 	}
-	defer it.Close()
 
-	tallyResult := group.DefaultTallyResult()
+	if proposal.Status == group.PROPOSAL_STATUS_WITHDRAWN || proposal.Status == group.PROPOSAL_STATUS_ABORTED {
+		return nil, sdkerrors.Wrapf(errors.ErrInvalid, "can't get the tally of a proposal with status %s", proposal.Status)
+	}
 
-	var vote group.Vote
-	for {
-		_, err = it.LoadNext(&vote)
-		if errors.ErrORMIteratorDone.Is(err) {
-			break
-		}
-		if err != nil {
-			return group.TallyResult{}, err
-		}
+	var policyInfo group.GroupPolicyInfo
+	if policyInfo, err = q.getGroupPolicyInfo(ctx, proposal.Address); err != nil {
+		return nil, sdkerrors.Wrap(err, "load group policy")
+	}
 
-		var member group.GroupMember
-		err := q.groupMemberTable.GetOne(ctx.KVStore(q.key), orm.PrimaryKey(&group.GroupMember{
-			GroupId: groupId,
-			Member:  &group.Member{Address: vote.Voter},
-		}), &member)
-
-		switch {
-		case sdkerrors.ErrNotFound.Is(err):
-			// If the member left the group after voting, then we simply skip the
-			// vote.
-			continue
-		case err != nil:
-			// For any other errors, we stop and return the error.
-			return group.TallyResult{}, err
-		}
+	// Ensure that group hasn't been modified since the proposal submission.
```

Comment:
why are we still checking this since tally will eventually be done at the end?

Reply:
Yes

Reply:
This endpoint is for voting period active proposals. Tally of voting period ended proposals will be there in proposal info.

Reply:
could you rephrase please? I don't get what you mean.
if the group has been modified, eg one member left, then we just ignore his vote if he voted before he left
so we don't need to check whether the group has been modified or not

Reply:
updated, thanks for clarifying!